Src/Support/Google.Apis.Auth/OAuth2/CredentialFactory.cs

```diff
@@ -0,0 +1,397 @@
+                JsonCredentialParameters.AuthorizedUserCredentialType => CreateUserCredentialFromParameters(credentialParameters, targetCredentialType),
+                JsonCredentialParameters.ServiceAccountCredentialType => CreateServiceAccountCredentialFromParameters(credentialParameters, targetCredentialType),
+                JsonCredentialParameters.ExternalAccountCredentialType => CreateExternalCredentialFromParameters(credentialParameters, targetCredentialType),
+                JsonCredentialParameters.ImpersonatedServiceAccountCredentialType => CreateImpersonatedServiceAccountCredentialFromParameters(credentialParameters, targetCredentialType),
+                JsonCredentialParameters.ExternalAccountAuthorizedUserCredentialType => CreateExternalAccountAuthorizedUserCredentialFromParameters(credentialParameters, targetCredentialType),
+                _ => throw new InvalidOperationException($"Error creating credential from JSON or JSON parameters. Unrecognized credential type {credentialParameters.Type}."),
+            };
+    }
+
+    /// <summary>
+    /// Validates actual type can be cast to target type.
+    /// </summary>
+    /// <param name="targetType"> The type to which we validate actualType can be cast to.</param>
+    /// <param name="actualType"> The type to validate can be cast to the targetType.</param>
+    /// <exception cref="InvalidOperationException">Thrown when targetType is not assignable from actual type.</exception>
+    static void CheckCompatibility(Type targetType, Type actualType)
+    {
+        if(!targetType.IsAssignableFrom(actualType))
+        {
+            throw new InvalidOperationException($"The credential type {actualType.FullName} is not compatible with the requested type {targetType.FullName}");
+        }
+    }
+
+    /// <summary>
+    /// Creates a user credential from JSON data.
+    /// </summary>
+    /// <param name="credentialParameters">The JSON credential data.</param>
+    /// <param name="targetType">The requested credential type.</param>
+    /// <returns>The created credential.</returns>
+    private static UserCredential CreateUserCredentialFromParameters(JsonCredentialParameters credentialParameters, Type targetType)
+    {
+        if (credentialParameters.Type != JsonCredentialParameters.AuthorizedUserCredentialType ||
+            string.IsNullOrEmpty(credentialParameters.ClientId) ||
+            string.IsNullOrEmpty(credentialParameters.ClientSecret))
+        {
+            throw new InvalidOperationException("JSON data does not represent a valid user credential.");
+        }
+        if (credentialParameters.UniverseDomain is not null && credentialParameters.UniverseDomain != GoogleAuthConsts.DefaultUniverseDomain)
+        {
+            throw new InvalidOperationException($"{nameof(UserCredential)} is not supported in universe domains other than {GoogleAuthConsts.DefaultUniverseDomain}");
+        }
+
+        CheckCompatibility(targetType, typeof(UserCredential));
```

Comment:
```suggestion
        }
        CheckCompatibility(targetType, typeof(UserCredential));
```
But also, there's a lot of blank lines after. I think this may be your editor configuration, but it makes the code look longer than it actually is.

(This all can be done in a separate PR).

Reply:
I made sure to remove the extra space introduced around check compatibility.